### DISCUSSION TITLE
Best practices for more complex queries on data objects

### DISCUSSION THREAD

**Discussion by VincentBibauw** [pimcore/pimcore#13731] [Q&A / Support] [answered]
Hi,

What's the best practice for doing more complex queries on data object. 
For example, I want the 7 most common titles for books.

Is this the right way to do this? 
What's the best practice? 

```php
$books
    ->setGroupBy('title')
    ->setLimit(7)
    ->onCreateQueryBuilder(
        function (QueryBuilder $queryBuilder) {
            $queryBuilder
                ->addSelect('COUNT(*) as titleCount')
                ->orderBy('titleCount', 'DESC')
            ;
        }
    );
```

**Comment by dpfaffenbauer** [comment, marked as answer]
you can use CoreShop index for more complex queries

**Reply by VincentBibauw** [reply]
Hi @dpfaffenbauer 

In case the project is not a webshop, adding CoreShop as a dependency is not ideal. 
I'm looking for the (best) way to do this in Pimcore.

**Reply by mattamon** [reply]
Even though it is not a webshop you could still use the Ecommerce Framework Bundle Index.
https://pimcore.com/docs/pimcore/current/Development_Documentation/E-Commerce_Framework/Index_Service/index.html

You can store specific data and reduce the complexity of your queries, which would be the better option.

For my projects I used the Doctrine QueryBuilder a lot as your example is showing.
I used it more like this and targeted the tables directly.
```php
$builder = DB::getConnection()->createQueryBuilder();
```

**Reply by dpfaffenbauer** [reply]
@VincentBibauw you can install coreshop/index-bundle independently

**Reply by VincentBibauw** [reply]
I think for my use case, using Doctrine will be the best solution. 
Thanks!
